### PR TITLE
Deprecate the old feature syntax from config.xml

### PR DESCRIPTION
**Pull request by gorkem**
Deprecates the feature syntax and starts using the new plugin tags for
defining plugins. Adjusts the tests for the new logic. The old feature
syntax is still read or removed but all new entries are done with the new
syntax.

### CODE REVIEW

**Comment by omefire** [comment]
With this new syntax, how do you specify variables ?

**Comment by gorkem** [comment]
@omefire see the changes on test-config.xml for examples of the new tag

**Comment by omefire** [comment]
ok, got it.

**Comment by omefire** [comment]
reviewing this.

File: cordova-lib/spec-cordova/ConfigParser.spec.js [hunk 1]
```diff
@@ -114,78 +114,94 @@ describe('config.xml parser', function () {
...
+                   'org.apache.cordova.justafeature',
+                   'org.apache.cordova.legacyfeature'
                ];
-               var list = cfg.getFeatureIdList();
+               var list = cfg.getPluginIdList();
                expect(list.length).toEqual(expectedList.length);
-               expectedList.forEach(function(feature){
-                   expect(list).toContain(feature);
+               expectedList.forEach(function(plugin){
+               expect(list).toContain(plugin);
```
Comment (omefire) [inline cordova-lib/spec-cordova/ConfigParser.spec.js]: :nit please indent this line.

File: cordova-lib/spec-cordova/ConfigParser.spec.js [hunk 2]
```diff
@@ -114,78 +114,94 @@ describe('config.xml parser', function () {
                 expect(cfg.getPlatformPreference('orientation', 'foobar')).toEqual('');
             });
         });
-        describe('feature',function(){
-            it('should read feature id list', function() {
+        describe('plugin',function(){
+            it('should read plugin id list', function() {
                var expectedList = [
                    'org.apache.cordova.featurewithvars',
```
Comment (omefire) [inline cordova-lib/spec-cordova/ConfigParser.spec.js]: since we are changing names from 'feature' to 'plugin', these ids should be updated. instead of 'org.apache.cordova.featurewithvars', please use 'org.apache.cordova.pluginwithvars'.

File: cordova-lib/spec-cordova/test-config.xml [hunk 3]
```diff
@@ -1,53 +1,34 @@
...
-        <param name="version" value="1.1.1" />
-    </feature>
-    <feature name="A feature with url and version">
-        <param name="id" value="org.apache.cordova.featurewithurlandversion" />
-        <param name="version" value="1.1.1" />
-        <param name="url" value="http://cordova.apache.org/featurewithurlandversion" />
-    </feature>
-    <feature name="A simple feature">
-        <param name="id" value="org.apache.cordova.justafeature" />
+    <plugin name="org.apache.cordova.featurewithvars">
```
Comment (omefire) [inline cordova-lib/spec-cordova/test-config.xml]: please, update from 'org.apache.cordova.featurewithvars' to 'org.apache.cordova.pluginwithvars'.

File: cordova-lib/spec-cordova/test-config.xml [hunk 4]
```diff
@@ -1,53 +1,34 @@
...
+        <variable name="var" value="varvalue" />
+    </plugin>
+    <plugin name="org.apache.cordova.featurewithurl" src="http://cordova.apache.org/featurewithurl" />
+    <plugin name="org.apache.cordova.featurewithversion" version="1.1.1" />
+    <plugin name="org.apache.cordova.featurewithurlandversion" src="http://cordova.apache.org/featurewithurlandversion" version="1.1.1" />
+    <plugin name="org.apache.cordova.justafeature" />
```
Comment (omefire) [inline cordova-lib/spec-cordova/test-config.xml]: how would we specify a plugin with local folder ? (as opposed to version and url).

File: cordova-lib/src/configparser/ConfigParser.js [hunk 5]
```diff
@@ -279,113 +269,108 @@ ConfigParser.prototype = {
...
+   /**
+    * Returns a list of plugin (IDs)
+    * @return {string[]} Array of plugin IDs
+    */
+    getPluginIdList: function () {
+        var plugins = this.doc.findall('plugin');
+        var result = plugins.map(function(plugin){
+            return plugin.attrib.name;
+        });
+        var features = this.doc.findall('feature');
```
Comment (omefire) [inline cordova-lib/src/configparser/ConfigParser.js]: A comment here to let future code readers would be great !
I can foresee people scratching their head as to if/why we are supporting both syntax.

File: cordova-lib/src/configparser/ConfigParser.js [hunk 6]
```diff
@@ -279,113 +269,108 @@ ConfigParser.prototype = {
...
-                // Invalid feature
-                continue;
-            }
-            id = idTag.attrib.value;
-            if (!!id) {
-                // Has id and id is non-empty
-                result.push(id);
-            }
+    addPlugin: function(attributes, variables){
+        if ( !attributes && !attributes.name ) return;
```
Comment (omefire) [inline cordova-lib/src/configparser/ConfigParser.js]: Why silently fail here ? shouldn't we throw instead ? the client of this module might think that everything is fine, when in fact, the changes he planned haven't gone through.

Comment (omefire) [inline cordova-lib/src/configparser/ConfigParser.js]: Other than the above comment, this method looks good to me !

File: cordova-lib/spec-cordova/ConfigParser.spec.js [hunk 7]
```diff
@@ -114,78 +114,94 @@ describe('config.xml parser', function () {
...
+                expect(plugin.version).toEqual('1.1.1');
+            });
+            it('should read plugin variables', function () {
+                var plugin = cfg.getPlugin('org.apache.cordova.featurewithvars');
+                expect(plugin.variables).toBeDefined();
+                expect(plugin.variables.var).toBeDefined();
+                expect(plugin.variables.var).toEqual('varvalue');
+            });
+            it('should allow adding a new plugin', function(){
+                cfg.addPlugin({name:'myfeature'});
```
Comment (omefire) [inline cordova-lib/spec-cordova/ConfigParser.spec.js]: This test does not seem to test all the way to the filesystem. 

You should use cfg.write() here to flush things to the filesystem, then reload from the filesystem using the constructor.
As it currently stands, we are just changing things in memory. But I suppose the goal is to test the end-to-end scenario: from memory to filesystem and back.

File: cordova-lib/spec-cordova/ConfigParser.spec.js [hunk 8]
```diff
@@ -114,78 +114,94 @@ describe('config.xml parser', function () {
...
+                var plugins = cfg.doc.findall('plugin');
+                var pluginNames = plugins.map(function(plugin){
+                    return plugin.attrib.name;
                 });
-                expect(featureNames).toContain('myfeature');
+                expect(pluginNames).toContain('myfeature');
             });
             it('should allow adding features with params', function(){
```
Comment (omefire) [inline cordova-lib/spec-cordova/ConfigParser.spec.js]: please, change 'features' to 'plugins', and 'params' to 'variables'.

File: cordova-lib/spec-cordova/ConfigParser.spec.js [hunk 9]
```diff
@@ -114,78 +114,94 @@ describe('config.xml parser', function () {
...
+                expect(plugin.variables.var).toEqual('varvalue');
+            });
+            it('should allow adding a new plugin', function(){
+                cfg.addPlugin({name:'myfeature'});
+                var plugins = cfg.doc.findall('plugin');
+                var pluginNames = plugins.map(function(plugin){
+                    return plugin.attrib.name;
                 });
-                expect(featureNames).toContain('myfeature');
+                expect(pluginNames).toContain('myfeature');
```
Comment (omefire) [inline cordova-lib/spec-cordova/ConfigParser.spec.js]: please, change 'myfeature' to 'myplugin'.

File: cordova-lib/src/configparser/ConfigParser.js [hunk 10]
```diff
@@ -279,113 +269,108 @@ ConfigParser.prototype = {
...
             return undefined;
         }
-        var feature = this.doc.find('./feature/param[@name="id"][@value="' + id + '"]/..');
-        if (null === feature) {
+        var pluginElement = this.doc.find('./plugin/[@name="' + id + '"]');
+        if (null === pluginElement) {
+            var legacyFeature =  this.doc.find('./feature/param[@name="id"][@value="' + id + '"]/..');
+            if(legacyFeature){
+                 var events = require('../events');
+                 events.emit('log', 'Found deprecated feature entry for ' + id +' in config.xml.');
```
Comment (omefire) [inline cordova-lib/src/configparser/ConfigParser.js]: should this be a warning ?

File: cordova-lib/spec-cordova/test-config.xml (hunk 4 above)
Comment (omefire) [inline cordova-lib/spec-cordova/test-config.xml]: never mind. we apparently use 'src' for both url and folder.

File: cordova-lib/src/configparser/ConfigParser.js [hunk 11]
```diff
@@ -279,113 +269,108 @@ ConfigParser.prototype = {
 
         return scriptElements.filter(filterScriptByHookType);
     },
-
+   /**
+    * Returns a list of plugin (IDs)
+    * @return {string[]} Array of plugin IDs
+    */
+    getPluginIdList: function () {
```
Comment (omefire) [inline cordova-lib/src/configparser/ConfigParser.js]: Nice !

File: cordova-lib/src/configparser/ConfigParser.js [hunk 12]
```diff
@@ -279,113 +269,108 @@ ConfigParser.prototype = {
...
-                el.append(p);
-            });
+    removePlugin: function(id){
+        if(id){
+            var theElement = this.doc.find('./plugin/[@name="' + id + '"]');
+            if(!theElement){
+                theElement = this.doc.find('./feature/param[@name="id"][@value="' + id + '"]/..');
+            }
+            if(theElement){
+                var childs = this.doc.getroot().getchildren();
```
Comment (omefire) [inline cordova-lib/src/configparser/ConfigParser.js]: childs -> children

File: cordova-lib/src/configparser/ConfigParser.js [hunk 13]
```diff
@@ -417,11 +402,11 @@ ConfigParser.prototype = {
     getEngines: function(){
```
Comment (omefire) [inline cordova-lib/src/configparser/ConfigParser.js]: Please, revert the changes to 'getEngines'. they screw up the indenting and they are a no-op.

File: cordova-lib/src/cordova/platform.js [hunk 14]
```diff
@@ -257,20 +257,20 @@ function getPlatformDetailsFromDir(dir, platformIfKnown){
     }
 
     return Q({
-	libDir: libDir,
-	platform: platform,
+    libDir: libDir,
```
Comment (omefire) [inline cordova-lib/src/cordova/platform.js]: Please, fix the indentation.

File: cordova-lib/src/cordova/platform.js [hunk 15]
```diff
@@ -290,16 +290,16 @@ function remove(hooksRunner, projectRoot, targets, opts) {
...
-	    targets.forEach(function(target) {
-		var platformName = target.split('@')[0];
-		var xml = cordova_util.projectConfig(projectRoot);
-		var cfg = new ConfigParser(xml);
-		events.emit('log', 'Removing ' + target + ' from config.xml file ...');
-		cfg.removeEngine(platformName);
-		cfg.write();
-	    });
-	}
+    if(opts.save || autosave){
```
Comment (omefire) [inline cordova-lib/src/cordova/platform.js]: Please, revert these changes. no-op.

File: cordova-lib/src/cordova/plugin.js [hunk 16]
```diff
@@ -308,8 +301,8 @@ module.exports = function plugin(command, targets, opts) {
 };
 
 function getVersionFromConfigFile(plugin, cfg){
-    var feature = cfg.getFeature(plugin); 
-    return feature && (feature.params.url || feature.params.installPath || feature.params.version);
+    var pluginEntry = cfg.getPlugin(plugin);
+    return pluginEntry && pluginEntry.version;
```
Comment (omefire) [inline cordova-lib/src/cordova/plugin.js]: This wouldn't work if config.xml contains a url or a folder.
This line should be changed to : return pluginEntry && pluginEntry.version || pluginEntry.src;

**Comment by gorkem** [comment]
merged
